src/encoder/mod.rs

```diff
@@ -0,0 +1,494 @@
+/// to encode tiff ifd directories with images.
+///
+/// See `DirectoryEncoder` and `ImageEncoder`.
+///
+/// # Examples
+/// ```
+/// # extern crate tempfile;
+/// # let mut file = tempfile::tempfile().unwrap();
+/// # let image_data = vec![0; 100*100*3];
+/// use tiff::encoder::*;
+///
+/// let mut tiff = TiffEncoder::new(&mut file).unwrap();
+///
+/// tiff.write_image::<colortype::RGB8>(100, 100, &image_data).unwrap();
+/// ```
+pub struct TiffEncoder<W> {
+    writer: TiffWriter<W>,
+}
+
+impl<W: Write + Seek> TiffEncoder<W> {
+    pub fn new(writer: W) -> TiffResult<TiffEncoder<W>> {
+        let mut encoder = TiffEncoder {
+            writer: TiffWriter::new(writer),
+        };
+        
+        NativeEndian::write_header(&mut encoder.writer)?;
+
+        Ok(encoder)
+    }
+
+    /// Create a `DirectoryEncoder` to encode an ifd directory.
+    pub fn new_directory(&mut self) -> TiffResult<DirectoryEncoder<W>> {
+        DirectoryEncoder::new(&mut self.writer)
+    }
+
+    /// Create an 'ImageEncoder' to encode an image one slice at a time.
+    pub fn new_image<C: ColorType>(&mut self, width: u32, height: u32) -> TiffResult<ImageEncoder<W, C>> {
+        let encoder = DirectoryEncoder::new(&mut self.writer)?;
+        ImageEncoder::new(encoder, width, height)
+    }
+
+    /// Convenience function to write an entire image from memory.
+    pub fn write_image<'a, C: ColorType>(&mut self, width: u32, height: u32, data: &'a [C::Inner]) -> TiffResult<()> where &'a [C::Inner]: TiffValue {
+        let encoder = DirectoryEncoder::new(&mut self.writer)?;
+        let mut image: ImageEncoder<W, C> = ImageEncoder::new(encoder, width, height).unwrap();
+
+        let mut idx = 0;
+        while image.next_strip_sample_count() > 0 {
+            let sample_count = image.next_strip_sample_count() as usize;
+            image.write_strip(&data[idx..idx+sample_count]).unwrap();
+            idx += sample_count;
+        }
+        image.finish()
+    }
+}
+
+/// Low level interface to encode ifd directories.
+///
+/// You should call `finish` on this when you are finished with it.
+/// Encoding can silently fail while this is dropping.
+pub struct DirectoryEncoder<'a, W: Write + Seek> {
+    writer: &'a mut TiffWriter<W>,
+    dropped: bool,
+    // We use BTreeMap to make sure tags are written in correct order
+    ifd_pointer_pos: u64,
+    ifd: BTreeMap<u16, (u16, u32, Vec<u8>)>,
+}
+
+impl<'a, W: Write + Seek> DirectoryEncoder<'a, W> {
+    fn new(writer: &'a mut TiffWriter<W>) -> TiffResult<DirectoryEncoder<'a, W>> {
+        writer.pad_word_boundary()?;
+        let ifd_pointer_pos = writer.offset();
+        writer.write_u32(0)?;
+        Ok(DirectoryEncoder {
+            writer,
+            dropped: false,
+            ifd_pointer_pos,
+            ifd: BTreeMap::new(),
+        })
+    }
+
+    /// Write a single ifd tag.
+    pub fn write_tag<T: TiffValue>(&mut self, tag: Tag, value: T) {
```

Comment:
Values could be passed by reference instead of value here to accomodate `T: ?Sized` and thus `impl TiffValue for [u8]`. Nothing in this method strictly requires the value property.

Reply:
It would be possible to get the best of both worlds by adding an additional, generic impl:

```
impl<'a, T: TiffValue + ?Sized> TiffValue for &'a TiffValue {
    ...
}
```

Then neither this methods nor its callsites need changes but can benefit from the impl on `[u8]` directly.